etc/systemd/system-generators/zfs-mount-generator.c

```diff
@@ -0,0 +1,1084 @@
+	const char *p_systemd_requiresmountsfor = strtok_r(NULL, "\t", &toktmp); /* $14 */
+	const char *p_systemd_before            = strtok_r(NULL, "\t", &toktmp); /* $15 */
+	const char *p_systemd_after             = strtok_r(NULL, "\t", &toktmp); /* $16 */
+	      char *p_systemd_wantedby          = strtok_r(NULL, "\t", &toktmp); /* $17 */
+	      char *p_systemd_requiredby        = strtok_r(NULL, "\t", &toktmp); /* $18 */
+	const char *p_systemd_nofail            = strtok_r(NULL, "\t", &toktmp); /* $19 */
+	const char *p_systemd_ignore            = strtok_r(NULL, "\t", &toktmp); /* $20 */
+	/* END CSTYLED */
+
+	const char *pool = dataset;
+	if ((toktmp = strchr(pool, '/')) != NULL)
+		pool = strndupa(pool, toktmp - pool);
+
+	if (p_systemd_ignore == NULL) {
+		fprintf(stderr, PROGNAME ": %s (%d): not enough tokens!\n",
+		    dataset, getpid());
+		return (1 << 1);
+	}
+
+	strncpy(argv0, dataset, strlen(argv0));
+
+	// Minimal pre-requisites to mount a ZFS dataset
+	// By ordering before zfs-mount.service, we avoid race conditions.
+	const char *after = "zfs-import.target";
+	const char *wants = "zfs-import.target";
+	const char *bindsto = NULL;
+	char *wantedby = NULL;
+	char *requiredby = NULL;
+	bool noauto = false;
+	bool wantedby_append = true;
+
+	// If the pool is already imported, zfs-import.target is not needed.
+	// This avoids a dependency loop on root-on-ZFS systems:
+	//   systemd-random-seed.service After (via RequiresMountsFor)
+	//   var-lib.mount After
+	//   zfs-import.target After
+	//   zfs-import-{cache,scan}.service After
+	//   cryptsetup.service After
+	//   systemd-random-seed.service
+	if (tfind(pool, &known_pools, STRCMP)) {
+		after = "";
+		wants = "";
+	}
+
+	if (strcmp(p_systemd_after, "-") == 0)
+		p_systemd_after = NULL;
+	if (strcmp(p_systemd_before, "-") == 0)
+		p_systemd_before = NULL;
+	if (strcmp(p_systemd_requires, "-") == 0)
+		p_systemd_requires = NULL;
+	if (strcmp(p_systemd_requiresmountsfor, "-") == 0)
+		p_systemd_requiresmountsfor = NULL;
+
+	if (strcmp(p_encroot, "-") != 0) {
+		char *keyloadunit =
+		    systemd_escape(p_encroot, "zfs-load-key-", ".service");
+
+		if (strcmp(dataset, p_encroot) == 0) {
+			const char *keymountdep = NULL;
+			bool is_prompt = false;
+
+			regmatch_t uri_matches[3];
+			if (regexec(&uri_regex, p_keyloc,
+			    sizeof (uri_matches) / sizeof (*uri_matches),
+			    uri_matches, 0) == 0) {
+				p_keyloc[uri_matches[1].rm_eo] = '\0';
+				p_keyloc[uri_matches[2].rm_eo] = '\0';
+				const char *scheme =
+				    &p_keyloc[uri_matches[1].rm_so];
+				const char *path =
+				    &p_keyloc[uri_matches[2].rm_so];
+
+				if (0)
+					// Add known keylocations that do *not*
+					// need mounts here! http:// would fall
+					// into this, for example!
+					printf("scheme: '%s'", scheme);
+				else
+					keymountdep = path;
+			} else {
+				if (strcmp(p_keyloc, "prompt") != 0)
+					fprintf(stderr, PROGNAME ": %s (%d): "
```

Comment:
because there are lots of children and you would like to be able to map them to the ones you saw in the debug output, yes, to a lesser extent, but this is tied to `/dev/kmsg`, and suffixing the name with the PID in syslogish format is a great point! Updated all stderr writers to include it.